### PR TITLE
clojure-lsp: migrate to GraalVM

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

clojure-lsp now supports GraalVM builds that makes the binary faster to start and also seems to reduce memory usage.

Also bump version, to ensure we are using the latest fixes for GraalVM builds.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/misc/clojure-lsp/default.nix [hunk 1]
```diff
@@ -1,31 +1,63 @@
...
+
+  buildPhase = with lib; ''
+    args=("-jar" "${src}"
+          "-H:Name=clojure-lsp"
+          ${optionalString stdenv.isDarwin ''"-H:-CheckToolchain"''}
+          "-J-Dclojure.compiler.direct-linking=true"
+          "-J-Dclojure.spec.skip-macros=true"
+          "-H:+ReportExceptionStackTraces"
+          "--enable-url-protocols=jar"
+          # "-H:+InlineBeforeAnalysis"
```
Comment (thiagokokada) [inline pkgs/development/tools/misc/clojure-lsp/default.nix:24]: To be enabled in GraalVM 21.0.0.

File: pkgs/development/tools/misc/clojure-lsp/default.nix [hunk 2]
```diff
@@ -1,31 +1,63 @@
...
 
   dontUnpack = true;
 
-  buildInputs = [ makeWrapper ];
+  buildInputs = [ graalvm11-ce ];
+
+  buildPhase = with lib; ''
+    args=("-jar" "${src}"
+          "-H:Name=clojure-lsp"
+          ${optionalString stdenv.isDarwin ''"-H:-CheckToolchain"''}
```
Comment (thiagokokada) [inline pkgs/development/tools/misc/clojure-lsp/default.nix:19]: Didn't test, just copied it from `babashka` derivation.

**Comment by thiagokokada** [comment]
@GrahamcOfBorg build clojure-lsp

**Comment by thiagokokada** [comment]
Converting it in draft since the tests doesn't pass.

@GrahamcOfBorg build clojure-lsp

File: pkgs/development/tools/misc/clojure-lsp/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/misc/clojure-lsp/default.nix:27]: Will I notice something different when building on darwin?

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/chog">https://termbin.com/chog</a></li>
  </ul>
</details>

**Review by ericdallo — APPROVED** [review]
LGTM, just need to check why tests are not passing

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/kxb1">https://termbin.com/kxb1</a></li>
  </ul>
</details>

**Comment by ericdallo** [comment]
Ok, the issue with the tests is that we are not using the same jar clojure-lsp use to build the native-image, I'll change the clojure-lsp release to provide that jar too

File: pkgs/development/tools/misc/clojure-lsp/default.nix (hunk 2 above)
Comment (thiagokokada) [reply pkgs/development/tools/misc/clojure-lsp/default.nix:27]: Don't know, I don't have a macOS to test.

**Comment by thiagokokada** [comment]
@GrahamcOfBorg build clojure-lsp

**Comment by thiagokokada** [comment]
@GrahamcOfBorg build clojure-lsp

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/rqm4">https://termbin.com/rqm4</a></li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>clojure-lsp: log <a href="https://termbin.com/0ayw">https://termbin.com/0ayw</a></li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Seems we will probably need to bump GraalVM to version 21.0.0 first.

**Comment by thiagokokada** [comment]
Commit f6490f2 is building clojure-lsp completely from source. It will not work since `clojure-lsp` needs GraalVM 21.0.0 to compile, but since it is not working right now I decided to push this refactoring.

**Comment by thiagokokada** [comment]
Should be ready to merge now.

@SuperSandro2000 @ericdallo, can you please review?

File: pkgs/development/tools/misc/clojure-lsp/default.nix [hunk 3]
```diff
@@ -1,31 +1,55 @@
...
-      --add-flags "-server" \
-      --add-flags "-jar $out/share/java/${pname}.jar"
+    install -Dm755 ./clojure-lsp $out/bin/clojure-lsp
+  '';
+
+  # TODO: tests are timing out, needs more investigation
+  doCheck = false;
+  checkPhase = ''
+    ${babashka}/bin/bb ${src}/integration-test/run-all.clj ./clojure-lsp
   '';
```
Comment (thiagokokada) [inline pkgs/development/tools/misc/clojure-lsp/default.nix:46]: Running those tests manually seems to work, so probably there is some other issue.

**Comment by thiagokokada** [comment]
@GrahamcOfBorg build clojure-lsp

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>clojure-lsp: <a href="https://termbin.com/cxua">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/"content" wasn't supplied./raw/index.html">pretty log </a></li>
  </ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
clojure-lsp:
<details>


`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:33:3:

```
   |
33 |   buildPhase = with lib; ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:44:3:

```
   |
44 |   checkPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:38:3:

```
   |
38 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `fetchurl`.
Near pkgs/development/tools/misc/clojure-lsp/default.nix:1:29:

```
  |
1 | { lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
  |                             ^
```
</details>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>clojure-lsp</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
clojure-lsp:
<details>


`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:33:3:

```
   |
33 |   buildPhase = with lib; ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:44:3:

```
   |
44 |   checkPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/development/tools/misc/clojure-lsp/default.nix:38:3:

```
   |
38 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
Unused argument: `fetchurl`.
Near pkgs/development/tools/misc/clojure-lsp/default.nix:1:29:

```
  |
1 | { lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
  |                             ^
```
</details>
</details>

**Comment by thiagokokada** [comment]
Fixed the issue with the tests, now it is working.

@ericdallo Can you review it please?

**Comment by thiagokokada** [comment]
@GrahamcOfBorg build clojure-lsp
@GrahamcOfBorg test clojure-lsp

**Review by ericdallo — APPROVED** [review]
Nice, thank you very much!

**Comment by thiagokokada** [comment]
@SuperSandro2000 Can we have this PR merged?

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111864` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>clojure-lsp</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
clojure-lsp:
<details>


Unused argument: `fetchurl`.
Near pkgs/development/tools/misc/clojure-lsp/default.nix:1:29:

```
  |
1 | { lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
  |                             ^
```
</details>
</details>

File: pkgs/development/tools/misc/clojure-lsp/default.nix [hunk 4]
```diff
@@ -1,31 +1,66 @@
-{ lib, stdenv, fetchurl, jre, makeWrapper }:
+{ lib, stdenv, callPackage, fetchurl, fetchFromGitHub, leiningen, openjdk11
```
Comment (SuperSandro2000) [inline pkgs/development/tools/misc/clojure-lsp/default.nix:1]:
```suggestion
{ lib, stdenv, callPackage, fetchFromGitHub, leiningen, openjdk11
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Review by ericdallo — APPROVED** [review]

**Comment by thiagokokada** [comment]
One last rebase since @r-ryantm created another bump PR.

Could we approve and merge this? @SuperSandro2000 @ericdallo ?

**Review by ericdallo — APPROVED** [review]

**Pull request merged**